src/java.base/share/classes/java/lang/StackTraceElement.java

```diff
@@ -355,27 +355,58 @@ public boolean isNativeMethod() {
      * @revised 9
      * @see    Throwable#printStackTrace()
      */
+    @Override
     public String toString() {
-        String s = "";
-        if (!dropClassLoaderName() && classLoaderName != null &&
-                !classLoaderName.isEmpty()) {
-            s += classLoaderName + "/";
+        StringBuilder sb = new StringBuilder();
+        try {
+            appendTo(sb);
+        } catch (IOException shouldNotHappen) {
+            throw new RuntimeException("Error building StackTraceElement string", shouldNotHappen);
         }
-        if (moduleName != null && !moduleName.isEmpty()) {
-            s += moduleName;
+        return sb.toString();
+    }
 
-            if (!dropModuleVersion() && moduleVersion != null &&
-                    !moduleVersion.isEmpty()) {
-                s += "@" + moduleVersion;
+    /**
+     * Appends the {@link #toString()} representation of this stack trace element to the provided appendable.
+     * @param dest The destination appendable
+     * @throws IOException If an I/O error occurs
+     */
+    private void appendTo(Appendable dest) throws IOException {
```

Comment:
Perhaps this could be package-private for reuse by `Throwable.printStackTrace`, avoiding the intermediate toString.

Reply:
Yes, I think the `PrintStreamOrWriter` could implement `Appendable` as both `PrintStream` and `PrintWriter` implement `Appendable`. 

https://github.com/openjdk/jdk/blob/348a0521e1cd602c4093955310f838cf4ce4daae/src/java.base/share/classes/java/lang/Throwable.java#L756

This would also potentially allow optimizing away the additional intermediate strings from prefixing the lines with `"\tat "` concatenated with the `StackTraceElement.toString()`

https://github.com/openjdk/jdk/blob/348a0521e1cd602c4093955310f838cf4ce4daae/src/java.base/share/classes/java/lang/Throwable.java#L686-L687

https://github.com/openjdk/jdk/blob/348a0521e1cd602c4093955310f838cf4ce4daae/src/java.base/share/classes/java/lang/Throwable.java#L724-L725

Reply:
Just making notes for the future: having `PrintStreamOrWriter implement Appendable` works, but slightly messy due to `append` methods throwing checked `IOException` so these would likely need to catch & ignore. Given `Throwable.printStackTrace` is typically on an error handling path, if the `PrintStreamOrWriter` error stream/writer is not functioning properly, I imagine one would want to know but may be prevented from knowing.